openssl/src/cms.rs

```diff
@@ -6,16 +6,59 @@
 //! Data accepted by this module will be smime type `enveloped-data`.
 
 use ffi;
-use foreign_types::{ForeignType, ForeignTypeRef};
+use foreign_types::{ForeignType, ForeignTypeRef, Opaque};
+use libc::c_void;
 use std::ptr;
 
+use asn1::Asn1Object;
 use bio::{MemBio, MemBioSlice};
 use error::ErrorStack;
+use hash::MessageDigest;
 use libc::c_uint;
 use pkey::{HasPrivate, PKeyRef};
-use stack::StackRef;
-use x509::{X509, X509Ref};
-use {cvt, cvt_p};
+use stack::{Stack, StackRef};
+use x509::store::X509Store;
+use x509::{X509Ref, X509};
+use {cvt, cvt_n, cvt_p};
+
+pub struct CmsSignerInfoRef(Opaque);
+
+impl ForeignTypeRef for CmsSignerInfoRef {
+    type CType = ffi::CMS_SignerInfo;
+}
+
+pub struct CmsSignerInfo(*mut ffi::CMS_SignerInfo);
+
+impl ForeignType for CmsSignerInfo {
+    type CType = ffi::CMS_SignerInfo;
+    type Ref = CmsSignerInfoRef;
+
+    unsafe fn from_ptr(ptr: *mut ffi::CMS_SignerInfo) -> CmsSignerInfo {
+        CmsSignerInfo(ptr)
+    }
+
+    fn as_ptr(&self) -> *mut ffi::CMS_SignerInfo {
+        self.0
+    }
+}
+
+impl CmsSignerInfo {
+    pub fn add_attr(&mut self, obj: &Asn1Object, data_type: i32, data: &[u8]) -> i32 {
```

Comment:
This should probably return a result.

Is data_type an arbitrary integer or actually an enum?

Reply:
I can't find any documentation for it, but it seems to be an opaque integer.